clusterloader2/pkg/measurement/common/service_creation_latency.go

```diff
@@ -0,0 +1,323 @@
+	return []measurement.Summary{summary}, nil
+}
+
+func (s *serviceCreationLatencyMeasurement) handleObject(oldObj, newObj interface{}) {
+	var oldService *corev1.Service
+	var newService *corev1.Service
+	var ok bool
+	oldService, ok = oldObj.(*corev1.Service)
+	if oldObj != nil && !ok {
+		klog.Errorf("%s: uncastable old object: %v", s, oldObj)
+		return
+	}
+	newService, ok = newObj.(*corev1.Service)
+	if newObj != nil && !ok {
+		klog.Errorf("%s: uncastable new object: %v", s, newObj)
+		return
+	}
+	if isEqual := oldService != nil &&
+		newService != nil &&
+		equality.Semantic.DeepEqual(oldService.Spec, newService.Spec) &&
+		equality.Semantic.DeepEqual(oldService.Status, newService.Status); isEqual {
+		return
+	}
+
+	// TODO(#680): Make it thread-safe.
+	if !s.isRunning {
+		return
+	}
+	if newObj == nil {
+		if err := s.deleteObject(oldService); err != nil {
+			klog.Errorf("%s: delete checker error: %v", s, err)
+		}
+		return
+	}
+	if err := s.updateObject(newService); err != nil {
+		klog.Errorf("%s: create checker error: %v", s, err)
+	}
+}
+
+func (s *serviceCreationLatencyMeasurement) deleteObject(svc *corev1.Service) error {
+	key, err := runtimeobjects.CreateMetaNamespaceKey(svc)
+	if err != nil {
+		return fmt.Errorf("meta key created error: %v", err)
+	}
+	s.pingCheckers.DeleteAndStop(key)
+	return nil
+}
+
+func (s *serviceCreationLatencyMeasurement) updateObject(svc *corev1.Service) error {
+	// This measurement only works for services with LoadBalancer type.
+	// Service of this type has to have port specified.
+	if len(svc.Spec.Ports) < 1 {
+		return nil
+	}
+	key, err := runtimeobjects.CreateMetaNamespaceKey(svc)
+	if err != nil {
+		return fmt.Errorf("meta key created error: %v", err)
+	}
+	if _, exists := s.creationTimes.Get(key, creatingPhase); !exists {
+		s.creationTimes.Set(key, creatingPhase, svc.CreationTimestamp.Time)
+	}
+	if len(svc.Status.LoadBalancer.Ingress) < 1 {
+		return nil
+	}
+	if _, exists := s.creationTimes.Get(key, ipAssigningPhase); exists {
+		return nil
+	}
+	s.creationTimes.Set(key, ipAssigningPhase, time.Now())
+	klog.Infof("%s: %s ipAssigned", s, key)
+
+	pc := &pingChecker{
+		callerName:    s.String(),
+		svc:           svc,
+		creationTimes: s.creationTimes,
+		stopCh:        make(chan struct{}),
+	}
+	pc.run()
+	s.pingCheckers.Add(key, pc)
+
+	return nil
+}
+
+type pingChecker struct {
+	callerName    string
+	svc           *corev1.Service
+	creationTimes *measurementutil.ObjectTransitionTimes
+	stopCh        chan struct{}
+}
+
+func (p *pingChecker) run() {
+	key, err := runtimeobjects.CreateMetaNamespaceKey(p.svc)
+	if err != nil {
+		klog.Errorf("%s: meta key created error: %v", p.callerName, err)
+		return
+	}
+	success := 0
+	for {
+		select {
+		case <-p.stopCh:
+			return
+		default:
+			if _, exists := p.creationTimes.Get(key, reachabilityPhase); exists {
+				return
+			}
+			// TODO(#679): Current implementation handles only load balancers.
+			_, err := execservice.RunCommand(
```

Comment:
How long this command takes? In general I would like to understand what we will report if e.g. service will be programmed and available immediately.

Reply:
It's quite fast, although it need to pass 10 times. There will always be slight delay.
However if there is a time change in service becoming available, the measurement will spot it.

Reply:
10 times if sth takes 50ms is fine, 10 times if it takes 500ms doesn't sounds fine.

At the very least I would like to see an issue opened with a TODO here that we should validate if calling RunCommand for every single ping isn't too expensive (e.g. maybe after first success, we should run second that will do ping 9 times or sth like that?). But TODO is fine.

Reply:
Done. https://github.com/kubernetes/perf-tests/issues/685